wit/store.wit

```diff
@@ -0,0 +1,71 @@
+/// A keyvalue interface that provides eventually consistent key-value operations.
+/// 
+/// Each of these operations acts on a single key-value pair.
+/// 
+/// The value in the key-value pair is defined as a `u8` byte array and the intention is that it is
+/// the common denominator for all data types defined by different key-value stores to handle data,
+/// ensuring compatibility between different key-value stores. Note: the clients will be expecting
+/// serialization/deserialization overhead to be handled by the key-value store. The value could be
+/// a serialized object from JSON, HTML or vendor-specific data types like AWS S3 objects.
+/// 
+/// Data consistency in a key value store refers to the guarantee that once a write operation
+/// completes, all subsequent read operations will return the value that was written.
+/// 
+/// Any implementation of this interface must have enough consistency to guarantee "reading your
+/// writes." In particular, this means that you should never get a value that is older than the one
+/// you wrote, but you MAY get a newer value if one was written around the same time. For example,
+/// if a value at a given key is A, and you write B, then immediately read, you should get B. If
+/// something else writes C in quick succession, then you may get C.
+interface store {
+    use types.{bucket, error};
+
+    /// The crud resource contains the common create, read, update, and delete operations for
+    /// key-value pairs in a key-value store
+    resource store {
+        /// Create a new crud resource, using the given bucket to access the data
+        open: static func(store: borrow<bucket>) -> store;
```

Comment:
If we include `get`, `set`, ... to `bucket` resource and then extends `bucket` resource in other interfaces ike atomic, do they also inherit`get` and `set`, ... from `bucket` in `store` interface? This may not be what we want since there might be host which only wants to provider `wasi:keyvalue/batch` interface implementation. 

Building on your idea, I think we could define these as

```wit
// store.wit
interface store {
  resource bucket {
    open: ...
  }
  
  extend bucket with {
    read: ...
    write: ...
  }
}
```


```wit
// atomic.wit

interface atomic {
  use store.{bucket};
  
  extend bucket with {
    increment: ...
  }
}

```

Reply:
I also have thought of these as extensions and not that KV stores might only implement atomics or batch, but didn't know there was a syntax for it (unless the `extend bucket` syntax was just pseudocode)

Reply:
`extend bucket` syntax is indeed pseudocode atm. 

Yeah I take a step back and agree that atomics and batch are extensions to the crud operations since I could not find any evidence to back my argument up.

Reply:
Yes, `extend ... with` is just speculative syntactic sugar atm.  However, there is a non-sugar syntax described in my [other comment](https://github.com/WebAssembly/wasi-keyvalue/pull/41#discussion_r1544936228) that in theory should work (it's in WIT.md), but it seems not to be implemented, but it's perhaps a quick-and-easy thing to fill in (just relaxing the grammar).

Reply:
btw, unrelated to the discussion here, i've observed a common pattern in the keyvalue interfaces is that an interface is mostly just a wrapper around a resource. 

e.g. 

```wit
interface store {
  resource bucket { 
    ...
  }
}
```

or 

```wit
interface atomic {
  resource atomic {
    ...
  }
}
```

where all the operations are encapsulated within the resource so that the interfaces have nothing but type definitions left. 

would it make sense to remove interface as a wrapper of a resource? I guess that will be a big change to the WIT semantics so just leave this thought here for feedback.

Reply:
In theory, yes, that's totally expressible in the Component Model and so it could be made expressible in WIT too.  I get the impression that, when the dust settles in this proposal, we'll want the `store` to have 2 items:
```wit
interface store {
  open: func(name: string) -> bucket;
  resource bucket { ... }
}
```
and the other interfaces will contain multiple functions that extend `bucket` with additional operations.  So the use case may go away inside `wasi:keyvalue`, but we can see if the pattern shows up somewhere else.

Reply:
Ok, so to summarize: We'd love to do this, but can't do this now, is that correct?